storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java

```diff
@@ -0,0 +1,644 @@
+    
+    protected final LocalState _localState;
+    protected final String _profileCmd;
+    protected volatile boolean _exitedEarly = false;
+
+    private class ProcessExitCallback implements ExitCodeCallback {
+        private final String _logPrefix;
+
+        public ProcessExitCallback(String logPrefix) {
+            _logPrefix = logPrefix;
+        }
+
+        @Override
+        public void call(int exitCode) {
+            LOG.info("{} exited with code: {}", _logPrefix, exitCode);
+            _exitedEarly = true;
+        }
+    }
+
+    //For testing purposes
+    public BasicContainer(AdvancedFSOps ops, int port, LocalAssignment assignment,
+            Map<String, Object> conf, Map<String, Object> topoConf, String supervisorId, 
+            ResourceIsolationInterface resourceIsolationManager, LocalState localState,
+            String profileCmd) throws IOException {
+        super(ops, port, assignment, conf, topoConf, supervisorId, resourceIsolationManager);
+        _localState = localState;
+        _profileCmd = profileCmd;
+    }
+    
+    public BasicContainer(int port, LocalAssignment assignment, Map<String, Object> conf, String supervisorId,
+            LocalState localState, ResourceIsolationInterface resourceIsolationManager, boolean recover)
+            throws IOException {
+        super(port, assignment, conf, supervisorId, resourceIsolationManager);
+        _localState = localState;
+
+        if (recover) {
+            synchronized (localState) {
+                String wid = null;
+                Map<String, Integer> workerToPort = localState.getApprovedWorkers();
+                for (Map.Entry<String, Integer> entry : workerToPort.entrySet()) {
+                    if (port == entry.getValue().intValue()) {
+                        wid = entry.getKey();
+                    }
+                }
+                if (wid == null) {
+                    throw new ContainerRecoveryException("Could not find worker id for " + port + " " + assignment);
+                }
+                LOG.info("Recovered Worker {}", wid);
+                _workerId = wid;
+            }
+        } else {
+            createNewWorkerId();
+        }
+
+        String stormHome = System.getProperty("storm.home");
+        _profileCmd = stormHome + Utils.FILE_PATH_SEPARATOR + "bin" + Utils.FILE_PATH_SEPARATOR
+                + conf.get(Config.WORKER_PROFILER_COMMAND);
+    }
+
+    public BasicContainer(String workerId, Map<String, Object> conf, String supervisorId,
+            ResourceIsolationInterface resourceIsolationManager) throws IOException {
+        super(-1, null, conf, supervisorId, resourceIsolationManager);
+        _localState = null;
+        _workerId = workerId;
+        _profileCmd = null;
+    }
+
+    /**
+     * Create a new worker ID for this process and store in in this object and
+     * in the local state.  Never call this if a worker is currently up and running.
```

Comment:
Could this check if it is called at the wrong time? For example by throwing an IllegalStateException if _workerId is not null?